### PR TITLE
[SPARK-20647][core] Port StorageTab to the new UI backend.

### PR DESCRIPTION
**Pull request by vanzin**
This required adding information about StreamBlockId to the store,
which is not available yet via the API. So an internal type was added
until there's a need to expose that information in the API.

The UI only lists RDDs that have cached partitions, and that information
wasn't being correctly captured in the listener, so that's also fixed,
along with some minor (internal) API adjustments so that the UI can
get the correct data.

Because of the way partitions are cached, some optimizations w.r.t. how
often the data is flushed to the store could not be applied to this code;
because of that, some different ways to make the code more performant
were added to the data structures tracking RDD blocks, with the goal of
avoiding expensive copies when lots of blocks are being updated.

Tested with existing and updated unit tests.

### CODE REVIEW

**Comment by vanzin** [comment]
For context:

* Project link: https://issues.apache.org/jira/browse/SPARK-18085
* Upcoming PRs that build on this code: https://github.com/vanzin/spark/pulls

**Comment by vanzin** [comment]
Note that unit tests will fail; they depend on changes being added in #19678. I'm posting this to speed up the review process.

File: core/src/test/scala/org/apache/spark/status/AppStatusListenerSuite.scala [hunk 1]
```diff
@@ -685,12 +700,55 @@ class AppStatusListenerSuite extends SparkFunSuite with BeforeAndAfter {
       assert(exec.info.diskUsed === 0L)
     }
 
+    // Add a block from a different RDD. Verify the executor is updated correctly and also that
+    // the distribution data for rdd1 is up to date.
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/status/AppStatusListenerSuite.scala]: you're only checking the distribution data for rdd2, right?

**Review by squito — APPROVED** [review]
lgtm (at least, as much I can say now till the other parts are merged in and tests pass etc.)

I still need to make sure I understand StreamBlocks better before I would merge, but I do think you've got the same behavior as before

**Comment by squito** [comment]
I figured out what is going on with StreamBlocks now, both before and after this change, so I'm OK with it

**Comment by squito** [comment]
merged to master

**Pull request closed without merging**